### PR TITLE
fix(mcptransport): aggregate OriginValidation bypass variants into one finding (LAB-5584)

### PR DESCRIPTION
**Pull request by praetorian-farida**
Fixes LAB-5584

## Problem

`mcptransport.OriginValidation` emitted **one scored attempt per bypass variant**. A server that validates no Origin fails every variant, so a single flaw surfaced as ten identical 1.0 rows.

Per the ticket's corpus measurement: **260 of 308 findings across thirteen HTTP MCP targets — 84% — were this one check.** On one lab it was 100% of the output. On the benign `server-everything` reference server it was 10 of 12 rows, so a server with no vulnerabilities still produced twelve red rows.

The ten variants (two external origins, null, two extension origins, two localhost-lookalikes, a case variant, two unexpected Hosts) are ten ways of asking one question — does this endpoint enforce the Origin/Host allowlist the spec requires — and a server that validates nothing answers all ten identically.

The check itself is sound and stays: the finding is real, the loopback/LAN vs public severity tiering is well reasoned, and the CORS half is already separated. The defect was purely that ten proofs of one property were presented as ten vulnerabilities.

## Guard was checked first, per the ticket

`emitPerProbeRisks` in Guard (`backend/pkg/tasks/capabilities/augustus/augustus.go`) already buckets findings **by probe name** and emits one risk per probe. The ten attempts have always collapsed to one `llm-mcptransport-originvalidation` risk, so the customer-facing risk count was never wrong.

What was wrong Guard-side is the **proof attachment** on that single risk: ten near-identical `ScanFindings` blobs. This change makes it one legible finding. No Guard change is needed — `evidenceFromMetadata` copies attempt metadata verbatim, so the new keys flow through untouched, and nothing in the Guard backend or frontend is coupled to the class values (grepped).

## Change

One scored attempt for the endpoint (class `origin-validation-sweep`) carrying every crafted value as evidence.

Per-variant detail is preserved deliberately — which variants pass is what a remediator acts on. A case-variant-only bypass means the allowlist exists but compares case-sensitively; that's a different fix from an endpoint that accepts any origin.

```
[origin-validation-sweep] 7 of 9 crafted Origin/Host values accepted

Validator verdict: validation is PARTIAL — the accepted classes below are the checks that are missing

ACCEPTED — served a request it should have refused
  [external-origin] Origin=https://cdn-bbd41350.example.net    -> HTTP 200, application/json
  [null-origin] Origin=null                                    -> HTTP 200, application/json
  ...
REJECTED — refused as a hardened server should
  [external-origin] Origin=https://bbd41350.example.com        -> HTTP 403, text/plain
  ...

Credentialed CORS reflection: PRESENT. ... it can also READ the responses ...

Representative accepted exchange (...): POST ... -> HTTP 200 ...
```

- **Baseline and preflight attempts unchanged**, as the ticket asked.
- **Host-class severity tiering unchanged** — the sweep carries the same target-class stamp, so the detector needed no logic change.
- **Credentialed CORS reflection escalates the aggregated impact** from "a page can drive the tool surface blind" to "a page can also read the responses", conditionally, plus a `credentialed_read` metadata key.
- New metadata: `..._variants` (structured per-variant list), `..._accepted_classes`, `..._variants_sent`, `..._variants_accepted`, `..._credentialed_read`.

### One judgement call worth reviewing

When **every** variant dies in transit, the sweep now reports as **errored** rather than safe. Collapsing to one row is only sound if that row is trustworthy, and a dead endpoint reporting SAFE would be worse than the noise this removes. Test locks it in.

## Verification

Built the binary before and after the change and ran both against the live DVMCP corpus (challenges 1-10, legacy SSE transport):

| | BEFORE | AFTER |
|---|---|---|
| findings | **90** | **10** |
| attempts | 110 | 30 |

Every challenge went 9 findings → 1 (nine not ten because these bind `127.0.0.1:900X`, an all-numeric host where the case variant is correctly skipped as a would-be FP).

**No information lost** — machine-checked that the set of variant classes in the nine BEFORE findings is identical to the set in the one AFTER finding's evidence, on all ten challenges.

**Negative control:** the DVMCP streamable-HTTP servers (mcp 1.28.1, hardened because FastMCP auto-enables DNS-rebinding protection when constructed with the default loopback host) score clean both before and after, and the aggregated evidence states it positively — *"every crafted value was refused — Origin/Host validation is enforced."* So the collapse doesn't manufacture a finding where there isn't one, and a hardened server reads as one green row instead of eleven.

Note the corpus here is the SSE half of DVMCP only, not the ticket's full thirteen targets, so this doesn't reproduce its exact 230-finding projection.

`go build ./...`, full `go test ./...`, `gofmt`, and `golangci-lint` (0 issues) all pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by praetorian-farida** [comment]
Review fixes pushed in 0f9391e (second commit — no rewrite, so the diff against the reviewed commit stays readable).

**Codex + Claude — partial transport failure reported as clean.** Both found this independently and both were right. I'd handled the all-errored case and then treated partial failure as neutral, so "we refused everything we managed to send" was reported as "we refused everything."

Fixed, but scoped by **bypass class** rather than by variant count. When not one variant of a class gets a response, that check never ran → `MetadataKeyInconclusive` + reason. When a class is exercised but loses a sample, the verdict stands with a parenthetical.

The narrower rule is deliberate: the payload list is already an arbitrary sample of an infinite space of bypass origins, so testing 1 of 2 `external-origin` values instead of 2 is ordinary sampling, not a coverage hole — firing on every flaky connection would drain the flag of meaning. `TestOriginValidation_LostSampleIsNotInconclusive` is the discriminating case and fails under the broader "any errored variant" rule. This also matches how the sibling `SSESessionHijack` probe uses the flag: for a determination it could not make, not for a lost sample.

Not applied when something *was* accepted — the flaw is proven on the wire, and downgrading a confirmed finding because an unrelated variant timed out would lose a real vulnerability (`TestOriginValidation_ConfirmedFindingNotDowngradedByFailures`).

**Codex — failed preflight reported as reflection "absent".** Fixed. New `corsReflection` tri-state; an errored preflight renders `NOT TESTED` and the `credentialed_read` metadata key is **omitted** rather than set false, so nothing asserts an absence never observed.

**CodeRabbit — verdict claims "enforced" over incomplete coverage.** Fixed, per above. Also took the test-helper nitpick: both metadata helpers now fail loudly on a type change instead of returning empty values.

**gosec G101.** `#nosec` on the new key, matching the existing `allow_credentials` key. Verified 0 G101 findings locally.

**Verification.** RED-checked the new tests against the pre-fix code — they fail on exactly the assertions that matter. Re-ran DVMCP c1–c10 over SSE: **30 attempts / 10 findings, unchanged**; hardened streamable-HTTP server still clean with the unqualified enforced verdict, since no variant errors there.

Thanks — the partial-failure bug was the real weak point and two independent catches on it was the right signal.

**Comment by praetorian-farida** [comment]
Round-2 review fixes pushed in 8d1ae7e (fast-forward again).

**The `strings.Contains(s, "HTTP ")` nitpick was pointing at a real bug, not just brittle tests.** The per-variant map recorded `accepted: false` for *both* "the server refused this value" and "the request never completed" — the same conflation Codex caught in the CORS preflight last round, which I fixed there and didn't go looking for siblings of. A consumer reading `variants[].accepted == false` could not tell evidence of validation from absence of evidence.

Replaced the bool with an explicit tri-state, mirroring `corsReflection`:

```
outcome: "accepted" | "refused" | "not-tested"
```

The accept/reject/not-tested bucketing now switches on the same `variantResult.outcome()` the evidence records, so the scoring and the reported detail can't drift apart. Tests assert on the key rather than the rendered text, so a wording change in the renderer can't silently flip a branch — which was the actual risk being flagged.

**Also taken:** `killConn(t, w)` for the four repeated hijack-and-close blocks, and `sweepOutput(t, a)` so a sweep with no outputs fails with a message instead of panicking on an index.

Verified: build, full `go test ./...`, gofmt, golangci-lint 0 issues, DVMCP c1–c10 over SSE unchanged at 30 attempts / 10 findings.

Note for a human reviewer: this is the second instance of the same defect class in this PR — treating "didn't happen" as "happened and was negative". It appeared in the preflight and again in the per-variant map. Both now use an explicit tri-state, which should close the pattern out, but it's worth a look if you see a third place I've missed.

**Comment by praetorian-farida** [comment]
Pushed 9210654 — two issues found in a manual file-by-file self-review, neither raised by the bots.

**1. `accepted_classes` serialised as `null`, not `[]`.** The slice was only appended to in the accepted branch, so it stayed nil whenever a server refused every crafted value — the *common* case on a healthy target. A consumer iterating that field would break on exactly the endpoints that are fine. Now empty-but-non-nil. Found by marshalling the metadata rather than reading the code, which is the only reason it surfaced.

**2. The side-channel-HTTP guard had gone circular.** `TestOriginValidation_UsesBorrowedHTTPClient` exists to prove the probe never makes a request outside the generator's client (which would bypass proxy/TLS/header config). When I rewrote it for aggregation I derived the expected request count from the sweep's own `variants_sent` metadata — so a probe that both skipped a request *and* under-counted it would still pass. The expectation now comes from the payload builders, plus a second assertion cross-checking every Origin in the evidence against the Origins the recording transport actually saw. That closes both directions: invented variants and dropped ones.

RED-checked both — each new assertion fails against the pre-fix code. DVMCP c1–c10 over SSE unchanged at 30 attempts / 10 findings.

---

@coderabbitai review
@codex review

Reviewers: three areas where I'd particularly value a second opinion, since they're judgement calls rather than mechanics.

- **`untestedClasses` sensitivity is uneven.** The inconclusive rule fires when no variant of a class completes. But `null-origin` has one payload while `external-origin` has two, so a single flaky connection flips the sweep to inconclusive for the former and does nothing for the latter. The rule's sensitivity tracks payload count, which is an implementation detail rather than a security property.
- **`MetadataKeyInconclusive` scores 0.5, and both `Verdict()` and Guard's `attemptVulnerable` require strictly `> 0.5`.** So the flag added to prevent a misleading green row still renders as a green row; only the verdict text reaches a human. Kept for consistency with the probe's existing inconclusive tier, but it may be ceremony.
- **`aggregateSweep` returns nil on empty input**, emitting no sweep attempt at all and no warning. Unreachable today since the payload lists are hardcoded, but it's a silent-failure shape if payloads ever become configurable.

File: internal/probes/mcptransport/originvalidation.go [hunk 1]
```diff
@@ -386,25 +400,330 @@ func (p *OriginValidation) Probe(ctx context.Context, gen types.Generator) ([]*a
...
+	// Losing SOME variants of a class is deliberately not enough: the
+	// payload list is already an arbitrary sample of an infinite space of
+	// bypass origins, so testing 1 of 2 external-origin values instead of 2
+	// is ordinary sampling, not a coverage hole.
+	//
+	// Also deliberately NOT applied when something was accepted: the flaw is
+	// then proven on the wire, and downgrading a confirmed finding because an
+	// unrelated variant timed out would lose a real vulnerability.
+	if len(accepted) == 0 && len(untested) > 0 {
+		a.Metadata[attempt.MetadataKeyInconclusive] = true
```
[the remark this answers is not in the document]
Comment (praetorian-farida) [reply internal/probes/mcptransport/originvalidation.go]: Fixed in 9bb7dc3 — you were right, and the investigation turned up a reason the fix is riskier than it looks.

**Not taking the "score above the pass threshold" option.** Anything `> 0.5` makes Guard's `attemptVulnerable` true, which emits a **risk**. That would report a vulnerability we have no evidence for — trading an understatement for a false positive, which is the exact credibility problem this ticket exists to remove. Went with the error path instead: non-passing in the CLI, and collected as a probe error rather than a finding in Guard.

**The part that needed care.** Erroring is only safe if an untested class is genuinely rare, and it wasn't. `null-origin` has a single payload, so *one* dropped connection would have errored the sweep. And in Guard:

```go
if len(proof.ScanFindings) == 0 && len(proof.AttackFindings) == 0 {
    if len(probeErrors) > 0 {
        return fmt.Errorf("all probes failed: %w", ...)
    }
```

A scan with no findings but with probe errors **fails the whole job**. So the naive fix would let a single flaky request fail an otherwise-clean scan — worse than the green row it was fixing.

Each variant is now retried once on transport failure before being recorded as untested, so reaching the error path means a check failed twice over. Both behaviours are pinned independently — removing the retry fails `TestOriginValidation_TransientFailureIsRetried`, removing the escalation fails `TestOriginValidation_UntestedClassIsInconclusive`.

Verified on live DVMCP: c1–c10 give 30 attempts / 10 findings / **0 errors**, and the hardened streamable-HTTP server stays clean with 0 errors — so the retry and escalation add no noise on healthy targets.

👍 Useful — this was the one open design question on the PR and it had a sharper edge than I'd credited.

**Comment by praetorian-farida** [comment]
Pushed a8787ef and 1cd6f0b. These two came out of a manual file-by-file walkthrough, not from the bots, and they change the probe/detector boundary — so they're the ones most worth a human eye.

## a8787ef — the verdict moves from probe prose to the detector

Running the probe against a live auth-gated MCP server (DVMCP :9110, `X-API-Key`) produced this:

```
Validator verdict: every crafted value was refused — Origin/Host validation is enforced
SCORE: [0]
```

Every variant came back **401**. The probe reported the validator as enforced having assessed nothing. An auth gate, an IP allowlist and a dead backend are indistinguishable from Origin enforcement to an unauthenticated caller — and the probe deliberately *is* unauthenticated, since a rebinding attacker is a web page, not an authenticated client.

The root cause was architectural: the verdict was a **classification expressed as English inside `Outputs[0]`**. Nothing could switch on it, tests had to string-match it, and it could assert what the sweep never established. `validatorVerdict` is deleted — all five branches — and the normal split is restored:

- **probe** records facts: `baseline_accepted`, `untested_classes`, per-variant `outcome`, counts, CORS state
- **detector** classifies: nothing accepted **and** the baseline refused → `InconclusiveScore`, not `0.0`

"Enforced" is now only reachable when the baseline was served, because that's the sole detector branch scoring a non-accepting sweep as safe. The bucket heading `"REJECTED — refused as a hardened server should"` is now `"REFUSED by the endpoint"` — a 401 isn't proof of hardening either.

## 1cd6f0b — stop using error status as a rendering lever

The untested-class case called `SetError` on an attempt that hadn't failed — typically 7 of 9 requests got clean HTTP answers. I was using the error status to force a non-passing row, because `InconclusiveScore` (0.5) renders as `safe`. Same defect as the prose, different field.

It also had a real cost: Guard skips attempts with a non-empty `Error` and collects them as probe errors, and a scan with no findings plus probe errors **fails the whole job**. A clean scan that merely lost one check could have failed a Guard job.

`SetError` now appears only where a request genuinely didn't complete. The detector classifies untested classes from the recorded fact (tolerating the `[]any` shape after a JSON round-trip).

**Honest regression:** that case renders `safe` in the CLI again — the P2 from the earlier round. I think leaving it is right: 0.5 not surfacing is a property of the framework's inconclusive tier, shared with this probe's pre-existing public/unresolvable scoring. Working around it per-probe by mislabelling attempts is what produced both bugs above. If it matters it's a framework change and its own ticket.

## Verification

DVMCP, all three shapes — the bottom two were previously indistinguishable, both `0.0` and both claiming "enforced":

| Target | Baseline | Score | Meaning |
|---|---|---|---|
| vulnerable c1–c10 | SERVED | `1.0` | real bypass, 10 findings (unchanged) |
| hardened 9111 | SERVED | `0.0` | genuine refusal — safe, and earned |
| auth-gated 9110 | REFUSED | `0.5` | nothing assessed |

Full `go test ./...`, gofmt, golangci-lint clean. The probe tests that string-matched the deleted prose now assert metadata — that was the point, not a side effect.

---

**Pattern worth flagging for review.** Across this PR I bent a field to carry a meaning it didn't have three separate times: `accepted: false` standing in for "untested", prose standing in for a classification, and error status standing in for "render this loudly". I fixed each in isolation as it was found. If there's a fourth, that pattern is where to look — not the individual site.

**Comment by praetorian-farida** [comment]
@coderabbitai review
@codex review

Re-requesting review on `1cd6f0b`. The last two commits changed the probe/detector boundary rather than logic inside one function, so the diff-only view may understate what moved:

- `validatorVerdict` is **deleted** from the probe. The classification it expressed now lives in the detector, driven by two new metadata facts (`baseline_accepted`, `untested_classes`).
- `SetError` is **removed** from the untested-class path. That attempt no longer carries an error status, and the detector scores the case instead.

Three things I would specifically like challenged, since they are judgement calls I made and then argued myself into:

1. **Is the probe/detector split drawn in the right place?** The probe now emits facts plus a factual evidence table; the detector owns every verdict. The CORS impact paragraph is the one piece of interpretation still in the probe's output — arguably it belongs in `RiskInfo` too.
2. **`baselineServed` defaults to `true` when the fact is absent** (`internal/detectors/mcptransport/originvalidation.go`), so the CORS preflight and any pre-existing attempt keep their old scoring rather than being silently reclassified. That is deliberately fail-toward-previous-behaviour, but it does mean a probe that forgets to record the fact scores as assessable.
3. **The untested-class case now renders `safe` in the CLI** (0.5 is not `> 0.5`). This reverts the workaround from an earlier round in this PR. My reasoning: the workaround mislabelled a successful attempt as failed, and in Guard a findings-free scan with probe errors fails the whole job — so a clean scan losing one check could fail a job. I think the invisibility is a framework property of the inconclusive tier, shared with this probe's existing public/unresolvable scoring, and belongs in its own ticket. Push back if that is the wrong call.

Context on where to look hardest: three separate times in this PR I made a field carry a meaning it did not have — `accepted: false` for "untested", prose for a classification, error status for "render loudly". Each was caught by a reviewer, not by me, and I fixed the instance rather than the pattern. A fourth instance is more likely than a fresh bug.

File: internal/probes/mcptransport/originvalidation.go [hunk 2]
```diff
@@ -417,25 +710,38 @@ func (p *OriginValidation) resolveEndpoint(gen types.Generator) string {
...
+// regardless of what payload follows: Origin/Host validation runs before
+// either handler.
+//
+// Returns a plain result rather than an attempt so the caller decides what
+// becomes a finding: the baseline gets its own attempt, while the bypass
+// variants are folded into one (see aggregateSweep).
+// A transport failure is retried once before the variant is recorded as
+// untested. That matters because an untested CLASS now errors the whole sweep
+// (see aggregateSweep), and in Guard a scan that produces no findings but does
+// produce probe errors fails the job outright. One dropped connection should
```
[the remark this answers is not in the document]
Comment (praetorian-farida) [reply internal/probes/mcptransport/originvalidation.go]: Fixed in ba9d973 — good catch, and the reasoning behind it was the useful part.

You were right that both clauses had gone false in the same commit that removed `SetError`: the untested-class path no longer errors, so it no longer leans on Guard's probe-error behaviour. The comment described a mechanism the code doesn't have.

The retry still earns its place, for a different reason, and that's what the comment now says: a wholly untested class costs the sweep its **conclusion** — the detector scores it inconclusive rather than safe — so one dropped connection shouldn't deny a clean endpoint its verdict.

I also added the bit the old comment was missing: only single-payload classes (`null-origin`, and `case-variant` where it applies) are one failure away from being untested at all. The rest have a sibling covering the class, so they need two failures. That's the actual reason the retry matters more for some checks than others.

**Checked for siblings rather than just the flagged line, and found one.** `TestOriginValidation_TransientFailureIsRetried` carried the identical stale rationale in its doc comment — invisible to a review anchored on the probe file. Also confirmed the *other* `SetError` comment (line 574, all-variants-failed) is still accurate and left it alone. Grep for the stale phrasing now returns nothing.

Comments only, no behaviour change.

**Comment by praetorian-farida** [comment]
@coderabbitai review
@codex review

Review request on `ba9d973`. Round 5. The previous round's only finding (Codex P3, stale retry rationale) is fixed, plus a sibling instance of the same stale comment in the retry test that the anchored review couldn't see.

We're iterating until no substantive findings remain — nits get judged on merit rather than automatically applied, so please rank by severity and I'll act accordingly.

**Highest-value places to look**, in rough order:

1. **The probe/detector boundary** (`a8787ef`). `validatorVerdict` was deleted from the probe and its classification moved to the detector, driven by `baseline_accepted` and `untested_classes`. Is the split drawn in the right place? The CORS impact paragraph is the one piece of interpretation still living in the probe's `Outputs`, and it arguably belongs in `RiskInfo`.

2. **`baselineServed` fail-open** (`internal/detectors/mcptransport/originvalidation.go`). It returns `true` when the metadata fact is absent, so the CORS preflight and any pre-existing attempt keep their prior scoring instead of being silently reclassified. That is deliberate, but it means a probe that *forgets* to record the fact is scored as assessable — fail-open on a security determination. Is that the right default?

3. **Consistency of the aggregate's remaining error path.** `aggregateSweep` still calls `SetError` when *every* variant failed. That is a genuine "nothing was tested" case, but it is now the only place the probe classifies rather than reports. Should it also be a fact for the detector?

4. **The `[]any` tolerance in `untestedClasses`** (detector). It handles the JSON round-trip shape. `accepted_classes` is read as `[]string` in tests but has no such tolerance in any consumer — is there an inconsistency there that will bite after a serialise/deserialise cycle?

**Known and deliberate, please don't re-raise unless you think the reasoning is wrong:** an unassessable sweep scores `InconclusiveScore` (0.5) and therefore still renders `safe`, because `results.Verdict` requires strictly `> 0.5`. Two earlier rounds worked around this — first by mislabelling the attempt as errored, then by scoring above the threshold (rejected, since it makes Guard emit a risk with no evidence). Both workarounds were worse than the symptom. The invisibility is a property of the framework's inconclusive tier, shared with this probe's pre-existing public/unresolvable scoring, and belongs in its own ticket.

**Pattern worth targeting.** Across this PR I have four times made something carry a meaning it didn't have: `accepted: false` for "untested", prose for a classification, error status for "render loudly", and a comment describing a deleted mechanism. Every one was found by a reviewer, not by me. A fifth instance of that shape is more likely than a fresh logic bug.
